### PR TITLE
feat: add meta image for social media sharing

### PR DESCRIPTION
**Pull request by dcshzj**
## Problem

Sharing our site links on social media doesn't show much, and there is no way to show an image.

## Solution

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
- [x] No - this PR is backwards compatible

**Features**:

- Add the meta image as part of the OpenGraph protocol as an optional field in the meta settings field.

**Improvements**:

- Extract out the functionality to get the meta description as a standalone function for easier readability.

## Tests

- [ ] Go to Studio and edit any page
- [ ] Go to the "Meta settings" tab
- [ ] Verify that the image attachment component is visible and the label is "Meta image".

NOTE: The image will NOT be uploaded even if you try to attach. This is a bug that will be tackled separately.

### CODE REVIEW

File: packages/components/src/engine/metadata.ts
Comment (seaerchin) [inline packages/components/src/engine/metadata.ts:98]: question: what's `openGraph` here and what does it do?

Comment (dcshzj) [reply packages/components/src/engine/metadata.ts:98]: Opengraph is the protocol for sharing links on social media (docs: https://ogp.me/). The object's shape is taken from here: https://nextjs.org/docs/app/api-reference/functions/generate-metadata#opengraph

File: .github/workflows/chromatic.yml
Comment (adriangohjw) [inline .github/workflows/chromatic.yml:44]: curious - why is there a need to add this?

File: packages/components/src/engine/metadata.ts
Comment (adriangohjw) [inline packages/components/src/engine/metadata.ts:1]: import and use `ISOMER_PAGE_LAYOUTS` instead of hardcoding e.g. `ISOMER_PAGE_LAYOUTS.Database`

File: packages/components/src/engine/metadata.ts
Comment (adriangohjw) [inline packages/components/src/engine/metadata.ts:36]: if `meta.image` is found, which means it's intentionally uploaded by user, should we prioritize that?

similar to `getMetaDescription` where we do
```ts
if (props.meta?.description) {
    return props.meta.description
  }
```

File: packages/components/src/engine/metadata.ts
Comment (adriangohjw) [inline packages/components/src/engine/metadata.ts:68]: suggestion - fallback to `site.logoUrl` instead of undefined?

File: packages/components/src/engine/metadata.ts
Comment (adriangohjw) [inline packages/components/src/engine/metadata.ts:64]: understand that this is defensive but will `props.site.url` be undefined?
plus can be misleading to call this `fullUrl` and potentially return `props.page.permalink`

**Review by adriangohjw — CHANGES_REQUESTED** [review]

File: .github/workflows/chromatic.yml
Comment (dcshzj) [reply .github/workflows/chromatic.yml:44]: Context: https://opengovproducts.slack.com/archives/C06R4DX966P/p1732607260729969?thread_ts=1732597220.921359&cid=C06R4DX966P

File: packages/components/src/engine/metadata.ts
Comment (dcshzj) [reply packages/components/src/engine/metadata.ts:1]: Updated in 252ce172e3223a985e958c0b12cc5f2fda117008.

File: packages/components/src/engine/metadata.ts
Comment (dcshzj) [reply packages/components/src/engine/metadata.ts]: Updated in 252ce172e3223a985e958c0b12cc5f2fda117008.

File: packages/components/src/engine/metadata.ts
Comment (dcshzj) [reply packages/components/src/engine/metadata.ts]: Updated in 252ce172e3223a985e958c0b12cc5f2fda117008 as `canonicalUrl` instead

File: packages/components/src/engine/metadata.ts
Comment (dcshzj) [reply packages/components/src/engine/metadata.ts]: Updated in 252ce172e3223a985e958c0b12cc5f2fda117008.

File: packages/components/src/engine/metadata.ts
Comment (adriangohjw) [inline packages/components/src/engine/metadata.ts:60]:
```suggestion
const getMetaImage = (props: IsomerPageSchemaType) => {
  if (props.meta?.image) {
    return props.meta.image
  }

  switch (props.layout) {
    case ISOMER_PAGE_LAYOUTS.Article:
      return props.page.image?.src
    case ISOMER_PAGE_LAYOUTS.Content:
    case ISOMER_PAGE_LAYOUTS.Database:
    case ISOMER_PAGE_LAYOUTS.Index:
    case ISOMER_PAGE_LAYOUTS.Collection:
      return undefined
    case ISOMER_PAGE_LAYOUTS.Homepage:
      return props.content.find((item) => item.type === "hero")?.backgroundUrl
    case ISOMER_PAGE_LAYOUTS.File:
    case ISOMER_PAGE_LAYOUTS.Link:
    case ISOMER_PAGE_LAYOUTS.Search:
    case ISOMER_PAGE_LAYOUTS.NotFound:
      // NOTE: These pages do not appear in search results, so we don't need to
      // provide a meta description
      return undefined
    default:
      const _: never = props
      return undefined
  }
}

```

Comment (dcshzj) [reply packages/components/src/engine/metadata.ts:60]: Hmm no, I see where you're coming from but we can't do this. The file and link layouts do not have support for meta images, so the attribute will not be present.

Actually in fact I think even for meta descriptions they should be removed.

Comment (adriangohjw) [reply packages/components/src/engine/metadata.ts:60]: if the attribute isn't present, won't `props.meta?.image` be false and skip through that check?

Comment (dcshzj) [reply packages/components/src/engine/metadata.ts:60]: It's typescript that will not allow it, because the attribute doesn't exist on the link and file layouts so there will be a type error.

**Review by adriangohjw — APPROVED** [review]

**Comment by dcshzj** [comment]
Thanks @adriangohjw @seaerchin for reviewing. Will put this PR on hold first while I do a refactor of the image uploading. @sehyunidaaa brought up a good point about the uploading not working as intended, so will do that first in a separate PR before this gets merged

**Comment by dcshzj** [comment]
Going to merge this first, the image uploading aspect **will not work** on Studio but we can deprioritise this temporarily as it is not user-facing.

**Pull request merged**
